### PR TITLE
Store images in ceph

### PR DESCRIPTION
**Pull request by rbradford**
Unify all storage on ceph - including volumes and images. As part of this the major changes needed are:

1. Add a new backend to the image service to upload the images into rbd images in ceph.
2. Enhance the ceph storage backend to snapshot and protect the rbd images used to store ciao images and expose the rbd clone functionality that lets you create an copy-on-write volume backed by the image to boot.
3. Modify ciao-controller to use the new functionality when creating storage based workloads.
4. Modify workload templates to use storage.

### CODE REVIEW

**Comment by rbradford** [comment]
@kaccardi can you do an early review pass? In particular i'd like to know if you want to see the "pre-work" in a separate PR (all the changes bar the last two, +- the workload changes.)

**Comment by markdryan** [comment]
Everything is working fine for me apart from ciao-cli image upload.  This is causing one of the BAT tests to fail

--- FAIL: TestUploadImage (3.47s)

Changes may be required to the current BAT tests to get this to work.

File: ciao-image/datastore/ceph.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+// Ceph implements the DataStore interface for Ceph RBD based storage
+type Ceph struct {
+	ImageTempDir string
+	BlockDriver  storage.CephDriver
+}
+
+// Write copies an image into the posix filesystem.
+// If the image already exists it will be overridden.
```
Comment (markdryan) [inline ciao-image/datastore/ceph.go]: Is this comment correct?

File: ciao-image/datastore/ceph.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+	image, err := os.Create(imagePath)
+	if err != nil {
+		return 0, fmt.Errorf("Error creating temporary image file: %v", err)
+	}
+
+	// TODO(rbradford): Is there a better way
+	buf := make([]byte, 1<<16)
+	res, err := io.CopyBuffer(image, body, buf)
+
```
Comment (markdryan) [inline ciao-image/datastore/ceph.go]: I think we're missing an

```
err = image.Close()
if err  != nil 
```

here

File: ciao-image/datastore/ceph.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+// Ceph implements the DataStore interface for Ceph RBD based storage
+type Ceph struct {
+	ImageTempDir string
+	BlockDriver  storage.CephDriver
+}
+
+// Write copies an image into the posix filesystem.
+// If the image already exists it will be overridden.
+func (c *Ceph) Write(ID string, body io.Reader) (int64, error) {
+	imagePath := path.Join(c.ImageTempDir, ID)
```
Comment (markdryan) [inline ciao-image/datastore/ceph.go]: You could use https://golang.org/pkg/io/ioutil/#TempFile.  Might save you one line of code.

File: ciao-image/datastore/ceph.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+	_, err = c.BlockDriver.CreateBlockDevice(ID, imagePath, 0)
+	if err != nil {
+		return 0, fmt.Errorf("Error creating block device: %v", err)
+	}
+
+	err = c.BlockDriver.CreateBlockDeviceSnapshot(ID, "ciao-image")
+	if err != nil {
+		c.BlockDriver.DeleteBlockDevice(ID)
+		return 0, fmt.Errorf("Unable to create snapshot: %v", err)
```
Comment (markdryan) [inline ciao-image/datastore/ceph.go]: Don't we leak imagePath here?  Maybe we should move the call to os.Remove up a little.

File: ciao-image/datastore/ceph.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	err = os.Remove(imagePath)
+	if err != nil {
+		c.BlockDriver.DeleteBlockDevice(ID)
+		return 0, fmt.Errorf("Error deleting temporary file: %v", err)
+	}
+
+	return res, nil
+}
+
+// Delete removes an image from the posix filesystem
```
Comment (markdryan) [inline ciao-image/datastore/ceph.go]: I think the comment might be incorrect.

File: ciao-storage/ceph.go [hunk 6]
```diff
@@ -32,31 +32,67 @@ type CephDriver struct {
 }
 
 // CreateBlockDevice will create a rbd image in the ceph cluster.
-func (d CephDriver) CreateBlockDevice(imagePath *string, size int) (BlockDevice, error) {
-	// generate a UUID to use for this image.
-	ID := uuid.Generate().String()
+func (d CephDriver) CreateBlockDevice(volumeUUID string, imagePath string, size int) (BlockDevice, error) {
+	if volumeUUID == "" {
+		volumeUUID = uuid.Generate().String()
+	}
```
Comment (markdryan) [inline ciao-storage/ceph.go]: We should probably check supplied, i.e., not "",  UUIDs are valid by calling the uuid.Parse function.  I'm just wondering what would happen if someone included a '@' in a UUID.

File: ciao-storage/ceph.go [hunk 7]
```diff
@@ -32,31 +32,67 @@ type CephDriver struct {
...
+	err := cmd.Run()
+	if err != nil {
+		return fmt.Errorf("Error when running: %v: %v", cmd.Args, err)
+	}
+
+	cmd = exec.Command("rbd", "--id", d.ID, "snap", "protect", volumeUUID+"@"+snapshotID)
+
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("Error when running: %v: %v", cmd.Args, err)
```
Comment (markdryan) [inline ciao-storage/ceph.go]: Should we try to delete the snapshot here if we can't protect it?

**Comment by markdryan** [comment]
BTW, my BAT failures were unrelated to this PR.  They seem to be problems I'm having in Single VM with the latest docker.  I'm working on a fix.

File: ciao-storage/ceph.go (hunk 7 above)
Comment (rbradford) [inline ciao-storage/ceph.go]: Done!

**Comment by rbradford** [comment]
@markdryan Hopefully all feedback addressed now. Thanks!

@kaccardi would you like to take a look?

File: ciao-image/datastore/ceph.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	defer os.Remove(image.Name())
+
+	// TODO(rbradford): Is there a better way
+	buf := make([]byte, 1<<16)
+	res, err := io.CopyBuffer(image, body, buf)
+	if err != nil {
+		return 0, fmt.Errorf("Error writing to temporary image file: %v", err)
+	}
+
+	err = image.Close()
```
Comment (markdryan) [inline ciao-image/datastore/ceph.go]: We still leak the image file descriptor here if io.CopyBuffer fails.

**Comment by markdryan** [comment]
@rbradford Just one outstanding comment on the code.  Also we should update the commit message of the PR.

File: ciao-image/datastore/ceph.go (hunk 8 above)
Comment (rbradford) [inline ciao-image/datastore/ceph.go]: Fixed

**Comment by rbradford** [comment]
Rebased and passed BAT on SingleVM:

rob@ciao-singlevm:~/go/src/github.com/01org/ciao/_release/bat$ go test ./...
ok  	github.com/01org/ciao/_release/bat/base_bat	238.979s
ok  	github.com/01org/ciao/_release/bat/image_bat	1.260s
ok  	github.com/01org/ciao/_release/bat/storage_bat	2.675s

**Pull request merged**